File: Aman/Tree/Average of Levels in Binary Tree.cpp
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double>res;
        queue<TreeNode*>q;
        q.push(root);
        while(!q.empty())
        {
            int n=q.size();
            double sum=0;
            for(int i=0;i<n;i++)
            {
                TreeNode* temp=q.front();
                q.pop();
                sum+=temp->val;
                if(temp->left)
                  q.push(temp->left);
                if(temp->right)
                  q.push(temp->right);
            }
            sum=sum/n;
            res.push_back(sum);
        }
        return res;
    }
};

File: Harish Kumar/Trees/Average of Levels in Binary Tree/Average of Levels in Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        
        List<Double> ans = new ArrayList<>();
        
        Queue<TreeNode> mq = new LinkedList<>();
        
        mq.add(root);
        
        while(mq.size() > 0)
        {
            long sum = 0;
            int level = 0;
            
            Queue<TreeNode> temp = new LinkedList<>();
            while(!mq.isEmpty())
            {
                TreeNode n = mq.remove();
                sum += n.val;
                level++;
            
                if(n.left != null)
                {
                    temp.add(n.left);
                }
                if(n.right != null)
                {
                    temp.add(n.right);
                }
            }
            
            mq = temp;
            ans.add(sum * 1.0 / level);
        }
        return ans;
    }
}
